docs/best-practices/host-custom-binary-cache-with-s3.md

```diff
@@ -0,0 +1,267 @@
+# Host Custom Binary Cache with S3 {#host-custom-binary-cache-with-s3}
+
+## TL;DR
+
+A guide on how to set up your own S3 nix binary cache using MinIO S3 server.
+
+## How Software Stored in Nix? {#how-software-stored-in-nix}
+
+Multiple versions of the same software package can be installed on a system, making it
+possible to satisfy various dependency chains at a time. This enables the installation of
+multiple packages that depend on the same third package, but on different versions of it.
+To achieve this, all packages are installed in the global nix store under `/nix/store/`
+and are then symlinked to the respective locations. To verify a package's uniqueness, its
+whole directory is hashed, and the hash put into the name of the package's main folder.
+Every software built from the same Nix expression that uses the same dependency software
+versions results in the same hash, no matter what system it was built on. If any of the
+dependency software versions were changed, this will result in a new hash for the final
+package.
+
+Using symlinks to install a package and link all the right dependencies to it also enables
+atomic updating. To make this clearer, let's think of an example where software X is
+installed in an older version and should be updated. Software X is installed in its very
+own directory in the global nix store and symlinked to the right directory, let's say
+`/usr/local/bin/`. When the update is triggered, the new version of X is installed into
```

Comment:
Packages will be linked to `/run/current-system/sw/`(NixOS) or `/etc/profiles/per-user/your-username/`(home-manager).

I am not sure whether it is appropriate to add this to here, just for reference.

Reply:
TBH, I'd prefer you directly propose your changes to my work to make our life easier :)